src/main/kotlin/com/exactpro/th2/infraoperator/spec/strategy/redeploy/tasks/RabbitMQGcTask.kt

```diff
@@ -0,0 +1,105 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.infraoperator.spec.strategy.redeploy.tasks
+
+import com.exactpro.th2.infraoperator.OperatorState
+import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.buildEstoreQueue
+import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.buildMstoreQueue
+import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.mq.RabbitMQContext
+import com.exactpro.th2.infraoperator.util.ExtractUtils
+import com.exactpro.th2.infraoperator.util.HelmReleaseUtils
+import com.rabbitmq.http.client.domain.QueueInfo
+import mu.KotlinLogging
+import java.io.IOException
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
+
+class RabbitMQGcTask(
+    private val retryDelay: Long
+) : Task {
+    private val lock = ReentrantLock()
+    private var deleteCandidates: Set<String> = emptySet()
+
+    override fun run(): Unit = lock.withLock {
+        deleteRedundantQueues()
+    }
+
+    override fun getName(): String = "RabbitMQGC"
+
+    override fun getRetryDelay(): Long = retryDelay
+
+    private fun deleteRedundantQueues() {
+        try {
+            val potentialGarbage = hashSetOf<String>()
+            RabbitMQContext.getTh2Queues().asSequence()
+                .map(QueueInfo::getName)
+                .forEach(potentialGarbage::add)
+
+            if (potentialGarbage.isEmpty()) {
+                deleteCandidates = emptySet()
+                return
+            }
+
+            val operatorState = OperatorState.INSTANCE
+            operatorState.namespaces.forEach { namespace ->
+                potentialGarbage.remove(buildEstoreQueue(namespace))
+                potentialGarbage.remove(buildMstoreQueue(namespace))
+            }
+
+            operatorState.getAllBoxResources().asSequence()
+                .flatMap { cr ->
+                    val namespace = ExtractUtils.extractNamespace(cr)
+                    val name = ExtractUtils.extractName(cr)
+                    val hr = operatorState.getHelmReleaseFromCache(name, namespace)
+                    sequence<String> {
+                        yieldAll(HelmReleaseUtils.extractQueues(hr.componentValuesSection))
+                        yield(buildEstoreQueue(namespace, name))
+                    }
+                }.forEach(potentialGarbage::remove)
+
+            if (potentialGarbage.isEmpty()) {
+                deleteCandidates = emptySet()
+                return
+            }
+
+            val channel = RabbitMQContext.getChannel()
+            val nextDeleteCandidates = potentialGarbage.minus(deleteCandidates).toMutableSet()
+            deleteCandidates.intersect(potentialGarbage).forEach { queue ->
```

Comment:
question: Could you please explain the idea here? I don't get why we need to have `deleteCandidates` and `nextDeleteCandidates`?

Reply:
This task is executed parallel the main engine. I do double check on two nearest iteration to avoid getting inconsistent 
operator state and do redundant delete.

Reply:
If I understand the idea correctly this does not save us from getting into a problem when we will delete candidates from `deleteCandidates` that were redeclared in parallel. There is no synchronization between queue declaration and deletion - I don't believe this is safe to do

Reply:
... you right. We can't grantee until state change and update RabbitMQ aren't one atomic operation.